velox/exec/Spill.cpp

```diff
@@ -310,6 +351,179 @@ SpillPartition::createOrderedReader(
   return std::make_unique<TreeOfLosers<SpillMergeStream>>(std::move(streams));
 }
 
+namespace {
+size_t estimateOutputBatchRows(
+    const std::vector<std::unique_ptr<SpillMergeStream>>& streams,
+    const vector_size_t maxRows,
+    const size_t maxBytes) {
+  size_t numEstimations{0};
+  int64_t estimateRowSizeSum{0};
+  for (const auto& stream : streams) {
+    const auto streamEstimateRowSize = stream->estimateRowSize();
+    if (streamEstimateRowSize.has_value()) {
+      ++numEstimations;
+      estimateRowSizeSum += streamEstimateRowSize.value();
+    }
+  }
+
+  if (numEstimations == 0) {
+    return maxRows;
+  }
+
+  const auto estimateRowSize =
+      std::max<vector_size_t>(1, estimateRowSizeSum / numEstimations);
+  return std::min<vector_size_t>(
+      std::max<vector_size_t>(1, maxBytes / estimateRowSize), maxRows);
+}
+
+// This contains batching parameters and various kinds of batching buffers that
+// are reused across multiple merging rounds.
+struct BatchingBufferParams {
+  BatchingBufferParams(
+      const vector_size_t _maxBatchRows,
+      const size_t _maxBatchBytes,
+      const TypePtr& type,
+      memory::MemoryPool* pool)
+      : maxBatchRows(_maxBatchRows), maxBatchBytes(_maxBatchBytes) {
+    rowVector = std::static_pointer_cast<RowVector>(
+        BaseVector::create(type, maxBatchRows, pool));
+    spillSources.resize(maxBatchRows);
+    spillSourceRows.resize(maxBatchRows);
+  }
```

Comment:
Shall we not have constructor for this Params? We can compose the struct by directly accessing its member variables.

Reply:
Then we should make it a class instead of struct.